app/src/main/java/io/homeassistant/companion/android/controls/ClimateControl.kt

```diff
@@ -31,6 +33,8 @@ object ClimateControl : HaControl {
         "heat_cool" to TemperatureControlTemplate.FLAG_MODE_HEAT_COOL,
         "off" to TemperatureControlTemplate.FLAG_MODE_OFF
     )
+    private var currentMode = "off"
+    private val supportedModes = ArrayList<String>()
```

Comment:
This does not work as you expect if you have more than one control with different values. Consider the following example: https://pl.kotl.in/SM3ioCSag

Because an `object` is a singleton, it will keep the latest value you set. This means that, for example, if you have 2 climate entities which are in `off` and `cool` mode, `currentMode` will always be `cool`. When you do not interact with the latest added entity, this breaks your "finding the next mode" logic as it'll use the state of the second entity to update the first one.

Reply:
Explored the options that I had here:

  1. Use HashMap to save values in `ClimateControl.kt` locally, based on their `entityId`, basically the same idea as before, just with a fix to the singleton behavior. This has the advantage that the internal state would be the same as the one presented on the screen, as when the visible state is updated the internal one is too.
  2. Change `performAction` to supply the `entity`, including it's state. This is in my opinion from an interface perspective the cleanest design. However, in the caller at `HaControlsProviderService::performControlAction()` this entity is not directly available. So either:
    1. This needs to be cached within the `HaControlsProviderService`
    2. This is fetched from the server right before sending the request
  Fetching for all entities is a really bad idea, as that would double the response time and adding an unnecessary request that would be thrown away most of the times anyways. Caching within `HaControlsProviderService` also seems like overkill to me.
  
For those reasons I went with the first option. Tested it with 2 devices and it worked as expected, making toggling independent of the state of the other devices.

One final note I have is that the logic would break with multi-server environments where device IDs are available in multiple servers

Reply:
I agree that the first option seems the best solution here. However, now that you've mentioned it this can't be ignored as the app supports multiple servers and specifically mixing controls from different servers:

> the logic would break with multi-server environments where device IDs are available in multiple servers

Suggestion for a fix: the `area` and `baseUrl` arguments for the `provideControlFeatures` function come from a class that holds more information about the control, including the `systemId` which is unique. In the interface, change [these last two arguments](https://github.com/home-assistant/android/blob/master/app/src/main/java/io/homeassistant/companion/android/controls/HaControl.kt#L87-L88) to a single `info` argument that passes the entire class?

Reply:
Ah shit, shouldn't have mentioned it :smile: 

Fix suggestion makes sense, had to only touch 12 source files to make that work :grimacing: 
However for most controls both arguments were not used and therefore only a signature change.

As I'm now also using the systemId as ID for the templates I had to adapt the call service part. My assumption was that `systemId` would be of the format `[SYSTEM.]AREA.ENTITYID` where at max 2 `.` are within the string and if there are 2 that means the `system` is present, otherwise there is only 1 `.` and the system is not present.

Therefore I'm getting the `AREA.ENTITYID` by splitting the first part off, if >= 2 `.` are there